### PR TITLE
[WIP] Refactor training loop

### PR DESCRIPTION
**Pull request by vmoens**

### CODE REVIEW

File: torchrl/objectives/ppo.py [hunk 1]
```diff
@@ -434,11 +434,14 @@ def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
...
-                f"and {log_weight.shape})"
-            )
+        if advantage.shape != log_weight.shape:
+            try:
+                log_weight = log_weight.view_as(advantage)
+            except Exception as err:
+                raise RuntimeError(
+                    f"advantage.shape and log_weight.shape do not match (got {advantage.shape} "
+                    f"and {log_weight.shape})"
+                ) from err
```
Comment (vmoens) [inline torchrl/objectives/ppo.py:444]: I'm not sure we want to keep these changes. Let me explain what usually happens:

We usually have an input tensor of observations and actions of shape `[*B, F]` and `[*B, A]` respectively.
We pass the observation through the model, get a distribution with params `[*B, P]`. Then we measure the probability of the action. This will give a tensor of shape `[*B]` because the last dim vanishes in `log_prob` (the probability of a multivariate event is still a scalar, so to say).
Hence in PPO we unsqueeze the last dimension to get `[*B, 1]` which is also the expected size of the reward, done state and value.

For some reason, after my refactoring I ended up with a log_prob of size `[*B, 1]`, hence the unsqueeze was producing a shape that was unwanted. But this may be a bug on my side

File: torchrl/collectors/collectors.py [hunk 2]
```diff
@@ -300,7 +300,7 @@ def update_policy_weights_(
         if policy_weights is not None:
             self.policy_weights.apply(lambda x: x.data).update_(policy_weights)
         elif self.get_weights_fn is not None:
-            self.policy.load_state_dict(self.get_weights_fn())
+            self.policy_weights.apply(lambda x: x.data).update_(self.get_weights_fn())
```
Comment (vmoens) [inline torchrl/collectors/collectors.py:303]: mergin main will solve this

File: examples/nanoChatGPT/config/train_rlhf.yaml [hunk 3]
```diff
@@ -1,5 +1,5 @@
 # training
-episode_length: 25
+episode_length: 10
```
Comment (vmoens) [inline examples/nanoChatGPT/config/train_rlhf.yaml:2]: we can reset this value to 25 if you think it is appropriate

File: examples/nanoChatGPT/data/shakespeare.py [hunk 4]
```diff
@@ -49,6 +47,19 @@ def __getitem__(self, idx):
             batch_size=[self.block_size],
         )
 
+    def __getitems__(self, idx):
```
Comment (vmoens) [inline examples/nanoChatGPT/data/shakespeare.py:50]: the collector will look for getitems. If found, it will use that instead. With this implementation, we don't need to stack.

File: examples/nanoChatGPT/env.py [hunk 5]
```diff
@@ -19,22 +19,24 @@
 @torch.no_grad()
 def _step(self, tensordict):
     self.step_num += 1
-    prompt = tensordict["prompt"]
+    prompt = tensordict.get("prompt")
```
Comment (vmoens) [inline examples/nanoChatGPT/env.py:22]: nit: I'm using get instead of `__getitem__`

File: examples/nanoChatGPT/env.py [hunk 6]
```diff
@@ -19,22 +19,24 @@
...
 
     # perform the action
-    action = tensordict["action"].squeeze(-1)
+    action = tensordict.get("action").squeeze(-1)
 
     # compute the reward
-    if self.step_num >= self.config["episode_length"]:
-        reward = self.reward_model(prompt).unsqueeze(-1)
```
Comment (vmoens) [inline examples/nanoChatGPT/env.py:29]: Why did we unsqueeze? We end up with a reward of shape `[B, 1, 1]` when we set the spec to be `[B, 1]`

File: examples/nanoChatGPT/env.py [hunk 7]
```diff
@@ -46,20 +48,15 @@ def _step(self, tensordict):
 @torch.no_grad()
 def _reset(self, tensordict):
     self.step_num = 0
-    if tensordict is None or tensordict.is_empty():
-        # if no tensordict is passed, we generate a single set of hyperparameters
-        # Otherwise, we assume that the input tensordict contains all the relevant
-        # parameters to get started.
-        tensordict = TensorDict({}, batch_size=self.config["batch_size"])
```
Comment (vmoens) [inline examples/nanoChatGPT/env.py:53]: we create tensordict but never use it really

File: examples/nanoChatGPT/env.py [hunk 8]
```diff
@@ -46,20 +48,15 @@ def _step(self, tensordict):
...
-
     batch = next(self.dataloader)
 
+    prompt = batch.prompt[:self.batch_size[0], -self.block_size :]
     out = TensorDict(
         {
-            "prompt": batch.prompt[:, -self.config["block_size"] :],
-            "done": torch.zeros((*batch.prompt.shape[:-1], 1, 1), dtype=torch.bool),
```
Comment (vmoens) [inline examples/nanoChatGPT/env.py:60]: why shape (1, 1) at the end?

File: examples/nanoChatGPT/models/actor_critic.py [hunk 9]
```diff
@@ -42,8 +42,13 @@ def __init__(self, base_model):
...
         )
-        value_head = TensorDictModule(
-            value_head, in_keys=["x"], out_keys=["state_value"]
+        value_head = TensorDictSequential(
+            TensorDictModule(
+                value_head, in_keys=["x"], out_keys=["state_value"]
+            ),
+            TensorDictModule(
+                lambda x: x[:, -1, :], in_keys=["state_value"], out_keys=["state_value"]
+            ),
```
Comment (vmoens) [inline examples/nanoChatGPT/models/actor_critic.py:51]: Tell me if I'm wrong but I feel that for the reward, we take the last element, but for the value we don't. Is this expected?

File: examples/nanoChatGPT/train_rlhf.py [hunk 10]
```diff
@@ -10,24 +9,49 @@
...
+from torchrl.envs import set_exploration_type, ExplorationType
 from utils import load_and_update_config
+import tqdm
+import csv
+import tiktoken
+from copy import deepcopy
 
 HERE = Path(__file__).parent
 
+def append_to_csv(number, filename):
```
Comment (vmoens) [inline examples/nanoChatGPT/train_rlhf.py:26]: a terrible way to log rewards on the fly

File: examples/nanoChatGPT/train_rlhf.py [hunk 11]
```diff
@@ -10,24 +9,49 @@
...
 
     # ######## INIT MODELS ########
     actor, critic, critic_head = init_actor_critic(config)
-
+    actor.eval()  # deactivate dropout on all modules
+    critic.eval()
+    
     reward_model, _ = init_reward_model(config)
+    reward_model.requires_grad_(False)
+    reward_model.eval()
```
Comment (vmoens) [inline examples/nanoChatGPT/train_rlhf.py:54]: all models should be put in eval, especially the reward model

File: examples/nanoChatGPT/train_rlhf.py [hunk 12]
```diff
@@ -10,24 +9,49 @@
...
     setup(config)
 
     # ######## INIT MODELS ########
     actor, critic, critic_head = init_actor_critic(config)
-
+    actor.eval()  # deactivate dropout on all modules
+    critic.eval()
+    
     reward_model, _ = init_reward_model(config)
+    reward_model.requires_grad_(False)
```
Comment (vmoens) [inline examples/nanoChatGPT/train_rlhf.py:53]: turn off any gradient computation from reward model

File: examples/nanoChatGPT/train_rlhf.py [hunk 13]
```diff
@@ -10,24 +9,49 @@
...
+    enc = tiktoken.get_encoding("gpt2")
+
     config = load_and_update_config("config/train_rlhf.yaml")
     setup(config)
 
     # ######## INIT MODELS ########
     actor, critic, critic_head = init_actor_critic(config)
-
+    actor.eval()  # deactivate dropout on all modules
+    critic.eval()
```
Comment (vmoens) [inline examples/nanoChatGPT/train_rlhf.py:50]: all models on eval mode

File: examples/nanoChatGPT/train_rlhf.py [hunk 14]
```diff
@@ -71,41 +99,98 @@ def forward(self, tensordict):
...
+    # a quick rollout to init the actor
+    with torch.no_grad():
+        td = env.rollout(3, actor)
+    del td
+
+    test_config = deepcopy(config)
+    test_config["batch_size"] = 1
+    test_config["episode_length"] = 50
+    train_loader_test, _ = get_dataloaders(test_config)
+    test_env = RLHFEnv(reward_model=reward_model, config=test_config, dataloader=train_loader_test)
```
Comment (vmoens) [inline examples/nanoChatGPT/train_rlhf.py:123]: I create a test env with a single element in the batch size

File: examples/nanoChatGPT/train_rlhf.py [hunk 15]
```diff
@@ -71,41 +99,98 @@ def forward(self, tensordict):
...
+
+    def test():
+        training = actor.training
+        actor.eval()
+        with set_exploration_type(ExplorationType.MODE), torch.no_grad():
```
Comment (vmoens) [inline examples/nanoChatGPT/train_rlhf.py:128]: we explicitly say we don't want to sample actions but take the most probable one

File: examples/nanoChatGPT/config/train_rlhf.yaml (hunk 3 above)
Comment (apbard) [reply examples/nanoChatGPT/config/train_rlhf.yaml:2]: 10 tokens is very short. In practice you would generate way more than 10 tokens (e.g. at most 10 words)

File: examples/nanoChatGPT/env.py (hunk 8 above)
Comment (apbard) [reply examples/nanoChatGPT/env.py:60]: we had issues with shapes

File: examples/nanoChatGPT/env.py [hunk 16]
```diff
@@ -46,20 +48,15 @@ def _step(self, tensordict):
...
     self.step_num = 0
-    if tensordict is None or tensordict.is_empty():
-        # if no tensordict is passed, we generate a single set of hyperparameters
-        # Otherwise, we assume that the input tensordict contains all the relevant
-        # parameters to get started.
-        tensordict = TensorDict({}, batch_size=self.config["batch_size"])
-
     batch = next(self.dataloader)
 
+    prompt = batch.prompt[:self.batch_size[0], -self.block_size :]
```
Comment (apbard) [inline examples/nanoChatGPT/env.py:53]: why we need this? `:self.batch_size[0]`

File: examples/nanoChatGPT/train_rlhf.py (hunk 10 above)
Comment (tcbegley) [reply examples/nanoChatGPT/train_rlhf.py:26]: Why do this instead of using`logging` or similar?

File: examples/nanoChatGPT/env.py (hunk 16 above)
Comment (vmoens) [reply examples/nanoChatGPT/env.py:54]: we can drop it
It was because I was using the same dataloader for the test env and the other, but now I have changed the config of the test so it's not needed anymore

File: examples/nanoChatGPT/env.py (hunk 8 above)
Comment (vmoens) [reply examples/nanoChatGPT/env.py:60]: It's better if we stay consistent IMO

Here is the recap of the workflow I understood:


Our "state" (the prompt) has shape [B, Tokens]
It is embedded in [B, Tokens, F] and then passed through the transformer which returns another [B, Tokens, F] tensor

- For the rewards, we take the last token ([:, -1, :]) of this final tensor [B, F] and get a single reward through a reward layer, hence we end up with [B, 1] shape
- For the value it's pretty much the same story.
- For the action, we have the [B, Tokens, F] tensor. We take the last token, and get the logits of shape [B, 1, L] (note that here we index with [-1] whereas for rewards and value we index with -1: I would suggest using -1 here too). Then out of this [B, 1, L] we sample an action which has size [B, 1], and we get a log_prob of size [B, 1] too. If we were using the -1 index these would be [B] and I personally would feel more comfortable with that.
- So our batch is 
```
TensorDict({
  "action": Tensor([B, 1]),
  "log_prob": Tensor([B, 1]),
  "reward": Tensor([B, 1]),
  "done": Tensor([B, 1]),
  "prompt": Tensor([B, Tokens]),
}, [B])
```
Add a `T` dimension after `B` for a rollout.

I believe that the reason why you were unsqueezing the reward and done was to get [B, 1, 1] like we eventually have for log_prob, but that's one level of unsqueezing too much.

The problem I was mentioning before regarding PPO for which I added a custom patch was that in PPO we unsqueeze the log_prob, but as I mentioned above I think this problem will vanish if we index the input to lm_head with -1 intead of [-1]: our log_prob will have size [B] and we'll be able to pass it to the PPO loss without trouble.


TLDR: let's try to change this line:
```
            logits = self.lm_head(x[:, [-1], :]) # note: using list [-1] to preserve the time dim
```
to
```
            logits = self.lm_head(x[:, -1, :]) # note: using list [-1] to preserve the time dim
```
and see what happens
